src/libstore/store-api.cc

```diff
@@ -998,106 +1078,39 @@ std::map<StorePath, StorePath> copyPaths(
 
     Activity act(*logger, lvlInfo, actCopyPaths, fmt("copying %d paths", missing.size()));
 
-    auto sorted = srcStore.topoSortPaths(missing);
-    std::reverse(sorted.begin(), sorted.end());
-
-    auto source = sinkToSource([&](Sink & sink) {
-        sink << sorted.size();
-        for (auto & storePath : sorted) {
-            auto srcUri = srcStore.getUri();
-            auto dstUri = dstStore.getUri();
-            auto storePathS = srcStore.printStorePath(storePath);
-            Activity act(*logger, lvlInfo, actCopyPath,
-                makeCopyPathMessage(srcUri, dstUri, storePathS),
-                {storePathS, srcUri, dstUri});
-            PushActivity pact(act.id);
-
-            auto info = srcStore.queryPathInfo(storePath);
-            info->write(sink, srcStore, 16);
-            srcStore.narFromPath(storePath, sink);
-        }
-    });
+    /* auto sorted = srcStore.topoSortPaths(missing); */
+    /* std::reverse(sorted.begin(), sorted.end()); */
+
+    /* auto source = sinkToSource([&](Sink & sink) { */
+    /*     sink << sorted.size(); */
+    /*     for (auto & storePath : sorted) { */
+    /*         auto srcUri = srcStore.getUri(); */
+    /*         auto dstUri = dstStore.getUri(); */
+    /*         auto storePathS = srcStore.printStorePath(storePath); */
+    /*         Activity act(*logger, lvlInfo, actCopyPath, */
+    /*             makeCopyPathMessage(srcUri, dstUri, storePathS), */
+    /*             {storePathS, srcUri, dstUri}); */
+    /*         PushActivity pact(act.id); */
+
+    /*         auto info = srcStore.queryPathInfo(storePath); */
+    /*         info->write(sink, srcStore, 16); */
+    /*         srcStore.narFromPath(storePath, sink); */
+    /*     } */
+    /* }); */
+
+    std::vector<std::pair<ValidPathInfo, std::unique_ptr<Source>>> pathsToCopy;
+
+    for (auto & missingPath : missing) {
+        auto info = srcStore.queryPathInfo(missingPath);
+        auto source = sinkToSource([&](Sink & sink) {
+            srcStore.narFromPath(missingPath, sink);
+        });
+        pathsToCopy.push_back(std::pair{*info, std::move(source)});
+    }
 
-    dstStore.addMultipleToStore(*source, repair, checkSigs);
+    dstStore.addMultipleToStore(pathsToCopy, act, repair, checkSigs);
 
     #if 0
-    std::atomic<size_t> nrDone{0};
```

Comment:
You can remove the `#if 0...#endif`.

Reply:
I am not sure we should remove this. When @edolstra did commented this out, we lost the ability to native `cp` for filesystem -> filesystem stores, and are forced to serialize and deserialize everything.

I think that is bad, and that functionality should be restored.

Overall, want something like "multiple dispatch" for `copyPaths`, where based on both the source and dest stores we can choose more efficient vs more general methods.